check explicitly for not None on field values

mikeedwards opened the pull request:
This is a really subtle weird error, but I was getting errors serializing an optional ImageField with no file associated with it--the errors kept insisting that there ought to be a file. I tried adding `allow_empty_file=True` and that works fine for GeoModelSerializer and the regular DRF ModelSerializer, but broke with GeoFeatureModelSerializer.

It turns out, in the GeoFeatureModelSerializer code, if I check _explicitly_ for whether the value is `None`, I can get GeoFeatureModelSerializer to work. My hunch is that ImageField, when empty, is a value that is equal to `None` (i.e. `<ImageFieldFile: None>`), but will not pass a merely "falsey" condition.  So I had to write it as;
`if value is not None:`
rather than
`if value:`

This is actually what DRF ModelSerializer itself does, more or less.
https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/serializers.py#L430

mikeedwards commented:
Yeah, looking at the Django source, it seems like the override of `__eq__` in `FieldFile ` is doing what I thought:

https://github.com/django/django/blob/master/django/db/models/fields/files.py#L27

nemesisdesign commented:
Hi @mikeedwards,

Thanks for contributing.

When there is a bug, we need a failing test, when possible. In this case it looks its possible to add your use case to the test suite. Would you be so kind to add it? If yes, be sure it fails before applying your fix.

mikeedwards commented:
Good point, sorry for the oversight. The fix was simple, but the case behind it is complex and needed better coverage. The test in my latest commit fails without the fix I added and passes with it, both for ImageField and FileField

I felt like I added a fair bit a new tooling, with an additional model, serializer and view. My sense was that I should segregate these cases with  `LocatedFile` and  `LocatedImage`, rather than putting more responsibility on the `Location` model. Hopefully, if there are issues with  `FileField` or `ImageField` in the future, other devs can use some of the bits I added to extend the tests.

nemesisdesign commented:
hey @mikeedwards, very good job, thank you.

I noticed that the Pillow requirement slows down the automated testing build a bit and it doesn't add too much value because there's already the file test which without your fix fails.

In order to keep the build lean and also because travis-ci.org is a free service and we should make conscientious use of its resources, i'm removing the Image model and its related code in order to remove the Pillow dependency.

I hope you won't hate me for this :-)

:+1:

mikeedwards commented:
Nope, no problem at all! Honestly, I felt a little bad having to add the `Pillow` dependency, especially since the `FileField` and `ImageField` cases are essentially identical.

Want me to push a new commit with the change, or do you have a commit for later?

nemesisdesign commented:
@mikeedwards have done myself already, didn't want to bother you too much

mikeedwards commented:
Great!
Are you planning on a new release in the near future? If not, I'll just pin my project's requirements.txt to this latest revision on the repo.
Thanks for this great library! It's really been a dream to code with and simplified my work in many, many ways.

nemesisdesign commented:
yes, i'll release soon after having reviewed and merged #57

nemesisdesign commented:
PS: if you work with GIS apps, it could be great if you could take a look at https://github.com/ninuxorg/nodeshot lots of improvements to this library come from that project :-)

mikeedwards commented:
Ah, that's a cool project! Hopefully they can also benefit from these commits. Thanks!

nemesisdesign commented:
very welcome
